File: php/src/Models/MessageDetail.php
<?php

declare(strict_types=1);

namespace TempMail\Models;

class MessageDetail extends Message
{
    public readonly bool $isHtml;
    public readonly string $contentType;
    public readonly string $bodyPreview;

    /**
     * @param array<array{name: string, type: string, size: int, url?: string}> $attachments
     * @param array<string, string> $headers
     * @param string[] $cc
     */
    public function __construct(
        string $id,
        string $sender,
        string $subject,
        \DateTimeImmutable $date,
        public readonly string $bodyText = '',
        public readonly string $bodyHtml = '',
        public readonly array $attachments = [],
        public readonly string $bodyPreviewInput = '',
        public readonly string $contentTypeInput = '',
        public readonly string $raw = '',
        public readonly array $headers = [],
        public readonly array $cc = [],
        public readonly string $replyTo = '',
        public readonly string $messageId = '',
        public readonly int $size = 0,
    ) {
        // Normalize: strip HTML → text if only HTML provided
        $hasHtml = trim($bodyHtml) !== '';
        $hasText = trim($bodyText) !== '';

        $resolvedText = $hasText ? $bodyText : ($hasHtml ? self::stripHtml($bodyHtml) : '');

        parent::__construct(
            $id, $sender, $subject, $date,
            preview: '',
            hasAttachments: count($attachments) > 0,
        );

        // Re-assign via late binding pattern (readonly workaround via init)
        $this->isHtml = $hasHtml;

        $this->contentType = match(true) {
            $hasHtml && $resolvedText !== '' => 'multipart/alternative',
            $hasHtml                          => 'text/html',
            default                           => ($contentTypeInput !== '' ? $contentTypeInput : 'text/plain'),
        };

        $this->bodyPreview = $bodyPreviewInput !== ''
            ? $bodyPreviewInput
            : mb_substr(trim($resolvedText), 0, 200);
    }

    /** Strip HTML tags to plain text. */
    public static function stripHtml(string $html): string
    {
        if ($html === '') {
            return '';
        }
        // Remove style/script blocks
        $s = preg_replace('#<(style|script)[^>]*>.*?</\1>#is', '', $html) ?? $html;
        // Replace block elements with newlines
        $s = preg_replace('#<(br\s*/?|/p|/div|/tr|/li|/h\d)>#i', "\n", $s) ?? $s;
        // Strip remaining tags
        $s = strip_tags($s);
        // Decode entities
        $s = html_entity_decode($s, ENT_QUOTES | ENT_HTML5, 'UTF-8');
        // Collapse blank lines
        $s = (string) preg_replace('/\n{3,}/', "\n\n", $s);
        return trim($s);
    }
}

File: csharp/TempMail/Models/MessageDetail.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace TempMail.Models;

public record MessageDetail : Message
{
    public string BodyText    { get; init; }
    public string BodyHtml    { get; init; }
    public string BodyPreview { get; init; }
    public string ContentType { get; init; }
    public string Raw         { get; init; }
    public Dictionary<string, string> Headers { get; init; }
    public List<string> Cc    { get; init; }
    public string ReplyTo     { get; init; }
    public string MessageIdHeader { get; init; }
    public int Size           { get; init; }
    public bool IsHtml        { get; init; }
    public List<JsonElement> Attachments { get; init; }

    public MessageDetail(
        string Id, string Sender, string Subject, DateTimeOffset Date,
        string bodyText = "", string bodyHtml = "",
        List<JsonElement>? attachments = null,
        string bodyPreview = "", string contentType = "",
        string raw = "", Dictionary<string, string>? headers = null,
        List<string>? cc = null, string replyTo = "",
        string messageIdHeader = "", int size = 0)
        : base(Id, Sender, Subject, Date, Preview: "", HasAttachments: (attachments?.Count ?? 0) > 0)
    {
        BodyHtml    = bodyHtml ?? "";
        Attachments = attachments ?? new List<JsonElement>();
        Raw         = raw ?? "";
        Headers     = headers != null ? new Dictionary<string, string>(headers) : new Dictionary<string, string>();
        Cc          = cc ?? new List<string>();
        ReplyTo     = replyTo ?? "";
        MessageIdHeader = messageIdHeader ?? "";
        Size        = size;

        // Normalize
        bool hasHtml = !string.IsNullOrWhiteSpace(bodyHtml);
        bool hasText = !string.IsNullOrWhiteSpace(bodyText);

        BodyText = hasText ? bodyText! : (hasHtml ? StripHtml(bodyHtml!) : "");
        IsHtml   = hasHtml;

        ContentType = hasHtml && BodyText.Length > 0 ? "multipart/alternative"
                    : hasHtml                        ? "text/html"
                    : !string.IsNullOrEmpty(contentType) ? contentType
                    : "text/plain";

        var preview = !string.IsNullOrEmpty(bodyPreview) ? bodyPreview
                    : BodyText.Length > 200 ? BodyText[..200].Trim()
                    : BodyText.Trim();
        BodyPreview = preview;

        if (!string.IsNullOrEmpty(MessageIdHeader) && !Headers.ContainsKey("Message-ID"))
            Headers["Message-ID"] = MessageIdHeader;
    }

    /// <summary>Strip HTML tags and decode entities to plain text.</summary>
    public static string StripHtml(string html)
    {
        if (string.IsNullOrEmpty(html)) return "";
        var s = Regex.Replace(html, @"<(style|script)[^>]*>[\s\S]*?</\1>", "", RegexOptions.IgnoreCase);
        s = Regex.Replace(s, @"<(br\s*/?|/p|/div|/tr|/li|/h\d)>", "\n", RegexOptions.IgnoreCase);
        s = Regex.Replace(s, "<[^>]+>", "");
        s = s.Replace("&amp;", "&").Replace("&lt;", "<").Replace("&gt;", ">")
             .Replace("&quot;", "\"").Replace("&#39;", "'").Replace("&nbsp;", " ");
        s = Regex.Replace(s, @"\n{3,}", "\n\n");
        return s.Trim();
    }
}
